pkg/api/types.go

```diff
@@ -2070,10 +2070,83 @@ const (
 	StrategicMergePatchType PatchType = "application/strategic-merge-patch+json"
 )
 
+// Type and constants for component type.
+type ComponentType string
+
+// ComponentType constants define the known types of components.
+// Each type may have multiple instances.
+// Each instance's name is prefixed by the type, with a random suffix. See SimpleNameGenerator.
+const (
+	ComponentAPIServer         ComponentType = "apiserver"
+	ComponentControllerManager ComponentType = "controller-manager"
+	ComponentScheduler         ComponentType = "scheduler" // TODO: move mesos-specific component type & validation to contrib
+)
+
+// ComponentSpec describes the attributes that a component is created with.
+type ComponentSpec struct {
+	// Type of the component
+	Type ComponentType `json:"type"`
+	// Address to reach the component (scheme, host, port, path). Address paths determined by component type.
+	Address string `json:"address,omitempty"`
+}
+
+type ComponentPhase string
+
+// These are the valid phases of a component.
+const (
+	// ComponentPending means the component is starting, but not yet fully functional.
+	ComponentPending ComponentPhase = "Pending"
```

Comment:
The NodeStatus comes with a list of NodeConditions, that I don't really understand. So I made the ComponentPhase more like a FSM to avoid that complexity. I'd love to understand how node's status and conditions work more, but it's kinda hard to reverse engineer.